### PR TITLE
Add support for axis keyword arg to numpy.argmax()

### PR DESCRIPTION
**Pull request by itamarst**
This adds support for the `axis` keyword argument to `np.argmax` and `np.ndarray.argmax`.

~I could not get arbitrary dimensions working, because `np.transpose()` only works (as far as I can tell) with literals. Suggestions are welcome.~

### CODE REVIEW

File: numba/core/ir_utils.py [hunk 1]
```diff
@@ -1303,7 +1303,7 @@ def find_single_branch(label):
 
 
 arr_math = ['min', 'max', 'sum', 'prod', 'mean', 'var', 'std',
-            'cumsum', 'cumprod', 'argmin', 'argmax', 'argsort',
+            'cumsum', 'cumprod', 'argmin', 'argsort',
```
Comment (stuartarchibald) [inline numba/core/ir_utils.py:1306]: This change is probably what's causing parfors (parallelisation) tests to fail. By removing this string expressions like `<array instance>.argmax()` can't be canonicalised to `np.argmax(<array instance>)` and as a result they don't get spotted by the parallel function replacement pass. Is this change needed to make the rest of the tests pass?

Comment (itamarst) [reply numba/core/ir_utils.py:1306]: Thanks for the quick response!

And yes, I was over-aggressive apparently detaching the `lower_builtin` code and shouldn't have touched this bit.

File: numba/np/arraymath.py [hunk 2]
```diff
@@ -704,75 +704,132 @@ def array_argmin_impl(arry):
     return impl_ret_untracked(context, builder, sig.return_type, res)
 
 
-@lower_builtin(np.argmax, types.Array)
-@lower_builtin("array.argmax", types.Array)
-def array_argmax(context, builder, sig, args):
-    ty = sig.args[0].dtype
+@register_jitable
```
Comment (itamarst) [inline numba/np/arraymath.py:707]: Note that these three functions (array_argmax_impl) are unchanged internally, they are just moved into top-level functions.

**Comment by itamarst** [comment]
Any chance someone can look at this? I have more functions I'd like to implement, but don't want to get ahead of myself if e.g. I'm doing it wrong.

File: numba/np/arraymath.py [hunk 3]
```diff
@@ -704,75 +704,132 @@ def array_argmin_impl(arry):
...
-                    max_idx = idx
-                idx += 1
-            return max_idx
-    res = context.compile_internal(builder, array_argmax_impl, sig, args)
-    return impl_ret_untracked(context, builder, sig.return_type, res)
+        def array_argmax_impl(arr, axis=None):
+            if arr.ndim > 3:
+                # Transposing currently only works with literals, so can't
+                # calculate new axis order for arbitrary array dimensions.
+                raise ValueError("More than 3 dimensions are not supported.")
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:793]: Please could you explain what this restriction is for? This sort of thing seems to work fine:
```python
from numba import njit
import numpy as np

@njit
def foo(a):
    return np.transpose(a, (3, 4, 2, 0, 1))

x = np.arange(np.prod([2, 3, 5, 7, 11])).reshape((11, 7, 5, 3, 2))

got = foo(x)
expected = foo.py_func(x)
np.testing.assert_allclose(got, expected)
```
is the difficulty here perhaps in dealing with tuple manipulation or maybe that the return type depends on some run time value?

File: numba/tests/test_array_reductions.py [hunk 4]
```diff
@@ -938,6 +938,57 @@ def a_variations():
...
+            ]
+            py_functions = [
+                lambda a, _axis=axis: np.argmax(a, axis=_axis)
+                for axis in axises
+            ]
+            c_functions = [
+                jit(nopython=True)(pyfunc) for pyfunc in py_functions
+            ]
+            for (pyfunc, cfunc) in zip(py_functions, c_functions):
+                self.assertPreciseEqual(pyfunc(arr), cfunc(arr))
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py:962]: You may wish to use the `.py_func` attribute on a jit wrapped function (it is a reference to the pure python function) to avoid this?

File: numba/tests/test_array_reductions.py [hunk 5]
```diff
@@ -938,6 +938,57 @@ def a_variations():
...
+
+        @jit(nopython=True)
+        def jitargmax(arr, axis):
+            return np.argmax(arr, axis)
+
+        def assert_raises(arr, axis):
+            with self.assertRaises(ValueError):
+                np.argmax(arr, axis)
+            with self.assertRaises(ValueError):
+                jitargmax(arr, axis)
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py:976]: The Numba code base tends to use this pattern to ensure that exceptions are being raised from the expected places.
https://github.com/numba/numba/blob/e314821f48bfc1678c9662584eef166fb9d5469c/numba/tests/test_array_reductions.py#L477-L482

File: numba/np/arraymath.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/np/arraymath.py:707]: thanks, comments like this help with review :)

File: numba/np/arraymath.py [hunk 6]
```diff
@@ -704,75 +704,132 @@ def array_argmin_impl(arry):
...
+            if axis < 0:
+                axis = arr.ndim + axis
+
+            if axis < 0 or axis >= arr.ndim:
+                raise ValueError("Axis is out of bounds")
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:798]: This needs an explicit unit test to make sure it is raising appropriately. Further, the type of `axis` ought to be checked to ensure it is of an appropriate type and if it is not, then raise (along with a test to ensure correctness).

File: numba/np/arraymath.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/np/arraymath.py:793]: Seems like there's something going on with method redirection, is that involved here too?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this is a great start, `axis` is hard to implement but often requested! I've given this an initial review, most of my comments are actually questions to understand the issues encountered a bit more. Once I've understood those some more I'll take another look. Thanks again!

**Comment by itamarst** [comment]
Many thanks, I will try to address comments next week.

File: numba/np/arraymath.py (hunk 6 above)
Comment (itamarst) [reply numba/np/arraymath.py:798]: This has a unit test, `test_argmax_axis_out_of_range`. I will add extra check for type.

File: numba/np/arraymath.py (hunk 3 above)
Comment (itamarst) [reply numba/np/arraymath.py:793]: Given the following code:

```python
# Make chosen axis the last axis:
            new_axis_order = list(range(arr.ndim))
            for i in range(axis):
                new_axis_order[i] = i
            for i in range(axis, arr.ndim - 1):
                new_axis_order[i] = i + 1
            new_axis_order[arr.ndim - 1] = axis
            transposed_arr = arr.transpose(new_axis_order)
```

I get the following error:

```
Traceback (most recent call last):
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
    constraint(typeinfer)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
    self.resolve(typeinfer, typevars, fnty)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 586, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1538, in resolve_call
    return self.context.resolve_function_type(fnty, pos_args, kw_args)
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 211, in resolve_function_type
    raise last_exception
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 194, in resolve_function_type
    res = self._resolve_user_function_type(func, args, kws)
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 246, in _resolve_user_function_type
    return func.get_call_type(self, args, kws)
  File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 324, in get_call_type
    failures.raise_error()
  File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 229, in raise_error
    raise errors.TypingError(self.format())
numba.core.errors.TypingError: No implementation of function Function(<function argmax at 0x7fffe8b84268>) found for signature:
 
 >>> argmax(array(int64, 2d, C), axis=Literal[int](0))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'array_argmax': File: numba/np/arraymath.py: Line 773.
    With argument(s): '(array(int64, 2d, C), axis=int64)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
   
   During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
   During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
   
   
   File "numba/np/arraymath.py", line 812:
           def array_argmax_impl(arr, axis=None):
               <source elided>
               new_axis_order[arr.ndim - 1] = axis
               transposed_arr = arr.transpose(new_axis_order)
               ^
   
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
           constraint(typeinfer)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
           self.resolve(typeinfer, typevars, fnty)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 615, in resolve
           raise TypingError(msg)
       numba.core.errors.TypingError: Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       The above exception was the direct cause of the following exception:
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 301, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 345, in apply
           sig = generic(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 599, in generic
           disp, new_args = self._get_impl(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 695, in _get_impl
           impl, args = self._build_impl(cache_key, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 756, in _build_impl
           disp_type.get_call_type(self.context, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 528, in get_call_type
           self.dispatcher.get_call_template(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 316, in get_call_template
           self.compile(tuple(args))
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
           cres = self._compiler.compile(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
           raise retval
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
           retval = self._compile_core(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
           pipeline_class=self.pipeline_class)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
           return pipeline.compile_extra(func)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
           return self._compile_bytecode()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
           return self._compile_core()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
           raise e
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
           pm.run(self.state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
           raise patched_exception
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
           self._runPass(idx, pass_inst, state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
           return func(*args, **kwargs)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
           mutated |= check(pss.run_pass, internal_state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
           mangled = func(compiler_state)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
           raise_errors=self._raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
           errs = infer.propagate(raise_errors=raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
           raise errors[0]
       numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
       Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       File "numba/np/arraymath.py", line 812:
               def array_argmax_impl(arr, axis=None):
                   <source elided>
                   new_axis_order[arr.ndim - 1] = axis
                   transposed_arr = arr.transpose(new_axis_order)
                   ^
  raised from /home/itamarst/Devel/numba/numba/core/typeinfer.py:1071

During: resolving callee type: Function(<function argmax at 0x7fffe8b84268>)
During: typing of call at /home/itamarst/Devel/numba/numba/tests/test_array_reductions.py (955)


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/itamarst/Devel/numba/numba/tests/test_array_reductions.py", line 962, in test_argmax_axis_1d_2d_3d
    self.assertPreciseEqual(cfunc.py_func(arr), cfunc(arr))
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 421, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 360, in error_rewrite
    raise e
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 373, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
    cres = self._compiler.compile(args, return_type)
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
    raise retval
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
    pipeline_class=self.pipeline_class)
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
    return self._compile_bytecode()
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
    return self._compile_core()
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
    raise e
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
    pm.run(self.state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
    raise patched_exception
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
    self._runPass(idx, pass_inst, state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
    mangled = func(compiler_state)
  File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
    raise_errors=self._raise_errors)
  File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
    errs = infer.propagate(raise_errors=raise_errors)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function argmax at 0x7fffe8b84268>) found for signature:
 
 >>> argmax(array(int64, 2d, C), axis=Literal[int](0))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'array_argmax': File: numba/np/arraymath.py: Line 773.
    With argument(s): '(array(int64, 2d, C), axis=int64)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
   
   During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
   During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
   
   
   File "numba/np/arraymath.py", line 812:
           def array_argmax_impl(arr, axis=None):
               <source elided>
               new_axis_order[arr.ndim - 1] = axis
               transposed_arr = arr.transpose(new_axis_order)
               ^
   
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
           constraint(typeinfer)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
           self.resolve(typeinfer, typevars, fnty)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 615, in resolve
           raise TypingError(msg)
       numba.core.errors.TypingError: Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       The above exception was the direct cause of the following exception:
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 301, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 345, in apply
           sig = generic(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 599, in generic
           disp, new_args = self._get_impl(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 695, in _get_impl
           impl, args = self._build_impl(cache_key, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 756, in _build_impl
           disp_type.get_call_type(self.context, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 528, in get_call_type
           self.dispatcher.get_call_template(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 316, in get_call_template
           self.compile(tuple(args))
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
           cres = self._compiler.compile(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
           raise retval
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
           retval = self._compile_core(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
           pipeline_class=self.pipeline_class)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
           return pipeline.compile_extra(func)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
           return self._compile_bytecode()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
           return self._compile_core()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
           raise e
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
           pm.run(self.state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
           raise patched_exception
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
           self._runPass(idx, pass_inst, state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
           return func(*args, **kwargs)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
           mutated |= check(pss.run_pass, internal_state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
           mangled = func(compiler_state)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
           raise_errors=self._raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
           errs = infer.propagate(raise_errors=raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
           raise errors[0]
       numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
       Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       File "numba/np/arraymath.py", line 812:
               def array_argmax_impl(arr, axis=None):
                   <source elided>
                   new_axis_order[arr.ndim - 1] = axis
                   transposed_arr = arr.transpose(new_axis_order)
                   ^
  raised from /home/itamarst/Devel/numba/numba/core/typeinfer.py:1071

During: resolving callee type: Function(<function argmax at 0x7fffe8b84268>)
During: typing of call at /home/itamarst/Devel/numba/numba/tests/test_array_reductions.py (955)


File "numba/tests/test_array_reductions.py", line 955:
    def test_argmax_axis_1d_2d_3d(self):
        <source elided>
            py_functions = [
                lambda a, _axis=axis: np.argmax(a, axis=_axis)
                ^
```

Comment (itamarst) [reply numba/np/arraymath.py:793]: I am not sure how to fix the above.

**Comment by itamarst** [comment]
Addressed most comments, the dimension-limitation one I am not sure how to approach, the error is mysterious to me.

File: numba/np/arraymath.py (hunk 3 above)
Comment (itamarst) [reply numba/np/arraymath.py:794]: Here is an isolated reproducer of the problem of trying to create the new axis order programmatically, instead of with hardcoding: https://gist.github.com/itamarst/ebdef3a06a97d409a91a55c818196447

Comment (stuartarchibald) [reply numba/np/arraymath.py:794]: Thanks for creating a reproducer. I think the underlying issue here is that `array.transpose` expects a tuple and not a list. I imagine that this is in part to help static analysis, tuples are immutable so it's safe for their type to carry their size. This means that a transpose can easily be verified as potentially valid if the tuple size is the same as the array dimensionality. In the case of providing a list, I think this is technically possible but there would need to be a run time check to ensure that the list is the right size.

Given the actual implementation of `argmax` presented herein, is `new_axis_order` a predictable mutation of `range(arr.ndim)` (I'm fairly sure it is)? In which case I think there could be a "trick" to support higher dimensionality.

Comment (itamarst) [reply numba/np/arraymath.py:794]: The logic is "take tuple(range(arr.ndim)) and move the given axis so it's last". So e.g. for ndim 3, axis 1:  `(0, 1, 2) -> (0, 2, 1)`. So it's predictable but it depends on the axis input.

Comment (stuartarchibald) [reply numba/np/arraymath.py:794]: I think this may permit the current restriction to be removed: https://github.com/stuartarchibald/numba/commit/c8584986b2fc059450b816614438f6d78b64acf2

**Comment by itamarst** [comment]
Thanks to @stuartarchibald this now supports more than 3 dimensions, should be able to remove "Waiting for author" label.

File: numba/np/arraymath.py [hunk 7]
```diff
@@ -704,75 +705,123 @@ def array_argmin_impl(arry):
...
-                    if numpy_version >= (1, 18):
-                        return idx
-                    else:
-                        idx += 1
-                        continue
-                if v > max_value:
-                    max_value = v
-                    max_idx = idx
+@register_jitable
+def array_argmax_impl_datetime(arry):
```
Comment (itamarst) [inline numba/np/arraymath.py:709]: Just a reminder in case there's a different reviewer: these three functions were just moved unchanged from inside the wrapping function.

File: numba/tests/test_array_reductions.py [hunk 8]
```diff
@@ -938,6 +938,68 @@ def a_variations():
...
+        arr1d = np.array([0, 20, 3, 4])
+        arr2d = np.arange(6).reshape(2, 3)
+        arr2d[0,1] += 100
+
+        arr4d = np.arange(120).reshape(2, 3, 4, 5) + 10
+        arr4d[0,1,1,2] += 100
+        arr4d[1,0,0,0] += 100
+
+        for arr in [arr1d, arr2d, arr4d]:
+            axises = list(range(arr.ndim)) + [
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py:951]:
```suggestion
            axes = list(range(arr.ndim)) + [
```

File: numba/tests/test_array_reductions.py [hunk 9]
```diff
@@ -938,6 +938,68 @@ def a_variations():
...
+                -(i+1) for i in range(arr.ndim)
+            ]
+            py_functions = [
+                lambda a, _axis=axis: np.argmax(a, axis=_axis)
+                for axis in axises
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py:956]:
```suggestion
                for axis in axes
```

File: numba/tests/test_array_reductions.py [hunk 10]
```diff
@@ -938,6 +938,68 @@ def a_variations():
...
                 check(a)
 
+    def test_argmax_axis_1d_2d_4d(self):
+        arr1d = np.array([0, 20, 3, 4])
+        arr2d = np.arange(6).reshape(2, 3)
+        arr2d[0,1] += 100
+
+        arr4d = np.arange(120).reshape(2, 3, 4, 5) + 10
+        arr4d[0,1,1,2] += 100
+        arr4d[1,0,0,0] += 100
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py:948]:
```suggestion
        arr4d[0, 1, 1, 2] += 100
        arr4d[1, 0, 0, 0] -= 51
```
to add a bit more variation

File: docs/source/reference/numpysupported.rst [hunk 11]
```diff
@@ -335,6 +335,8 @@ The following methods of Numpy arrays are supported:
 * :meth:`~numpy.ndarray.view` (only the 1-argument form)
 * :meth:`~numpy.ndarray.__contains__` 
 
+Where applicable, the corresponding top-level Numpy functions (such as
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:338]:
```suggestion
Where applicable, the corresponding top-level NumPy functions (such as
```

**Review by stuartarchibald — COMMENTED** [review]
@itamarst Thanks for the update. Couple of minor stylistic things to resolve else looks good.

@sklam / @esc can you check c858498 please? I wrote the patch, it's designed to create an appropriate tuple for use in `transpose` as discussed here: https://github.com/numba/numba/pull/6928#discussion_r645731038. Thanks.

**Comment by esc** [comment]
@stuartarchibald I reviewed 858498 and can confirm, that it looks good!

**Comment by stuartarchibald** [comment]
> @stuartarchibald I reviewed 858498 and can confirm, that it looks good!

Presume c858498 is meant? thanks for checking @esc .

**Comment by itamarst** [comment]
I applied the minor fixes, so assuming everything passes, sounds like this is ready to merge?

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
@itamarst Congratulations on your first contribution to Numba!

**Comment by itamarst** [comment]
Thanks for the help! Next up on my list is adding axis to argmin, and then escalating difficulty until the final boss, "it'd be nice if Numba did einsum", which is frankly terrifying.

**Comment by esc** [comment]
Thanks for the patch!
